include/mxnet/ndarray.h

```diff
@@ -312,8 +514,25 @@ class NDArray {
    * This is an internal function used by system that normal user should not use
    */
   inline void CheckAndAlloc() const {
+    CHECK_EQ(storage_type(), kDefaultStorage);
     ptr_->CheckAndAlloc();
   }
+  /* !
+   * \brief Alloc memory for non-default storage
+   * aux_shape is only known at run time
+   */
+  inline void CheckAndAlloc(const std::vector<TShape> &aux_shapes) const {
+    CHECK_NE(storage_type(), kDefaultStorage);
+    ptr_->CheckAndAlloc(shape_, aux_shapes, dtype_);
+  }
+  inline void CheckAndAllocData(const TShape &storage_shape) const {
+    CHECK_NE(storage_type(), kDefaultStorage);
+    ptr_->CheckAndAllocData(storage_shape, dtype_);
+  }
+  inline void CheckAndAllocAuxData(size_t i, const TShape &aux_shape) const {
```

Comment:
In some case the sparse operator doesn't know how much memory to allocate at once. e.g. https://github.com/eric-haibin-lin/mxnet/blob/master/src/operator/tensor/elemwise_unary_op.h#L308 
we have to allocate some temp memory to work out the aux_shape of the result. This makes it easier to reuse allocated temp memory explicitly, but makes code harder to read.
Another way is to just request for a temp memory to just calculate the aux_shape, and leave efficient memory management to the storage manager instead.